corehq/util/view_utils.py

```diff
@@ -62,49 +62,48 @@ def inner(request, *args, **kwargs):
             return response
 
         except http.Http404 as e:
-            blob = json.dumps({
-                'error': 404,
-                'message': unicode(e),
-            })
-            logger.warning('Not found: %s', request.path,
-                           extra={
-                               'status_code': 404,
-                               'request': request,
-                           })
-            return http.HttpResponseNotFound(blob, content_type=JSON)
+            return _get_json_exception_response(404, request, e, log_message='Not found: %s')
         except PermissionDenied as e:
-            logger.warning(
-                'Forbidden (Permission denied): %s', request.path,
-                extra={
-                    'status_code': 403,
-                    'request': request,
-                })
-            blob = json.dumps({
-                'error': 403,
-                'message': unicode(e),
-            })
-            return http.HttpResponseForbidden(blob, content_type=JSON)
+            return _get_json_exception_response(403, request, e, log_message='Forbidden (Permission denied): %s')
         except BadRequest as e:
-            blob = json.dumps({
-                'error': 400,
-                'message': unicode(e),
-            })
-            return http.HttpResponseBadRequest(blob, content_type=JSON)
+            return _get_json_exception_response(400, request, e)
         except Exception as e:
-            data = {
-                'error': 500,
-                'message': unicode(e)
-            }
-            if settings.DEBUG:
-                data['traceback'] = traceback.format_exc()
-            return http.HttpResponse(
-                status=500,
-                content=json.dumps(data),
-                content_type=JSON
-            )
+            return _get_json_exception_response(500, request, e)
     return inner
 
 
+def _get_json_exception_response(code, request, exception, log_message=None):
+    if log_message:
+        logger.warning(
+            log_message, request.path,
+            extra={
+                'status_code': code,
+                'request': request,
+            })
+
+    data = _json_exception_response_data(code, exception)
+
+    return http.HttpResponse(
+        status=code,
+        content=json.dumps(data),
+        content_type=JSON
+    )
+
+
+def _json_exception_response_data(code, exception):
+    if isinstance(exception.message, unicode):
+        message = unicode(exception)
+    else:
+        message = str(exception).decode('utf-8')
+    data = {
+        'error': code,
+        'message': message
+    }
+    if code == 500 and settings.DEBUG:
+        data['traceback'] = traceback.format_exc()
```

Comment:
what I meant to say is if another exception is raised and then subsequently caught by a separate exception handler
```py
import traceback
dx = {}
try:
    dx[1]
except KeyError:
    try:
        dx.not_here
    except AttributeError:
        pass
    traceback.print_exc()
```
output:
```
Traceback (most recent call last):
  File "<string>", line 7, in <module>
AttributeError: 'dict' object has no attribute 'not_here'
```
I think it will be fine in this specific context because these are private functions used in a very specific context, but generally it's unsafe to call `traceback.format_exc()` because it references global state. It was easier to reason about when `traceback.format_exc()` was coded directly in the `except` block, but having it in a helper function makes it harder to quickly see if you'll get the intended result. For example, adding a "harmless" function call that raises and catches an exception inside the `except` block before this function is called would make this return the wrong result. A safer way to do this is to use `traceback.format_exception(etype, value, tb)`, but python2 makes it a bit annoying to get the `tb` object for a given exception: gotta use `sys.exc_info()` as the first statement in the `except` block and then pass the result of that into this function.

No need to change. Just pointing it out.